From f2cb5149d6e17acb3561631c4f0ca425dfa3478e Mon Sep 17 00:00:00 2001
From: Iker Pedrosa <ipedrosa@redhat.com>
Date: Wed, 9 Jul 2025 17:54:48 +0000
Subject: [PATCH] Add inline string formatting functions and malloc attribute

Add new inline string formatting functions pam_asprintf() and pam_snprintf(),
along with a pam_sprintf() macro wrapper. These provide safer alternatives to
the standard printf family functions with additional compile-time checks.

Commit: https://gitlab.com/redhat/centos-stream/rpms/pam/-/commit/6a608c05dc381aaed0bec62a96a3c01ca59fed2f
Backported-by: Sai Harsha <ssuryad@amazon.com>
---
 libpam/include/pam_cc_compat.h |  6 ++++++
 libpam/include/pam_inline.h    | 36 ++++++++++++++++++++++++++++++++++
 2 files changed, 42 insertions(+)

diff --git a/libpam/include/pam_cc_compat.h b/libpam/include/pam_cc_compat.h
index 1aaed77..f448356 100644
--- a/libpam/include/pam_cc_compat.h
+++ b/libpam/include/pam_cc_compat.h
@@ -15,6 +15,12 @@
 # define PAM_CLANG_PREREQ(maj, min)	0
 #endif
 
+#if PAM_GNUC_PREREQ(3, 0)
+# define PAM_ATTRIBUTE_MALLOC		__attribute__((__malloc__))
+#else
+# define PAM_ATTRIBUTE_MALLOC		/* empty */
+#endif
+
 #if PAM_GNUC_PREREQ(4, 6)
 # define DIAG_PUSH_IGNORE_CAST_QUAL					\
 	_Pragma("GCC diagnostic push");					\
diff --git a/libpam/include/pam_inline.h b/libpam/include/pam_inline.h
index 2e3a8d6..064c671 100644
--- a/libpam/include/pam_inline.h
+++ b/libpam/include/pam_inline.h
@@ -9,6 +9,9 @@
 #define PAM_INLINE_H
 
 #include "pam_cc_compat.h"
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
 
 /*
@@ -49,4 +52,37 @@ pam_str_skip_prefix_len(const char *str, const char *prefix, size_t prefix_len)
 #define pam_str_skip_prefix(str_, prefix_)	\
 	pam_str_skip_prefix_len((str_), (prefix_), sizeof(prefix_) - 1 + PAM_MUST_BE_ARRAY(prefix_))
 
+static inline char * PAM_FORMAT((printf, 1, 2)) PAM_NONNULL((1)) PAM_ATTRIBUTE_MALLOC
+pam_asprintf(const char *fmt, ...)
+{
+	int rc;
+	char *res;
+	va_list ap;
+
+	va_start(ap, fmt);
+	rc = vasprintf(&res, fmt, ap);
+	va_end(ap);
+
+	return rc < 0 ? NULL : res;
+}
+
+static inline int PAM_FORMAT((printf, 3, 4)) PAM_NONNULL((3))
+pam_snprintf(char *str, size_t size, const char *fmt, ...)
+{
+	int rc;
+	va_list ap;
+
+	va_start(ap, fmt);
+	rc = vsnprintf(str, size, fmt, ap);
+	va_end(ap);
+
+	if (rc < 0 || (unsigned int) rc >= size)
+		return -1;
+	return rc;
+}
+
+#define pam_sprintf(str_, fmt_, ...)						\
+	pam_snprintf((str_), sizeof(str_) + PAM_MUST_BE_ARRAY(str_), (fmt_),	\
+		     ##__VA_ARGS__)
+
 #endif /* PAM_INLINE_H */
-- 
2.47.1

